docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,244 @@
+# Shoot Control Plane Migration
+
+## Motivation
+
+Currently moving the control plane of a shoot cluster can only be done manually and requires deep knowledge of how exactly to transfer the resources and state from one seed to another. This can make it slow and prone to errors.
+
+Automatic migration can be very useful in a couple of scenarios:
+- Seed goes down and can't be repaired (fast enough or at all) and it's control planes need to be brought to another seed
+- Seed needs to be changed, but this operation requires the recreation of the seed (e.g. turn a single-AZ seed into a multi-AZ seed)
+- Seeds need to be rebalanced
+- New seeds become available in a region closer to/in the region of the workers and the control plane should be moved there to improve latency
+- Gardener ring, which is a self-supporting setup/underlay for a highly available (usually cross-region) Gardener deployment
+
+
+## Goals
+
+* Provide a mechanism to migrate the control plane of a shoot cluster from one seed to another
```

Comment:
> Do the two seeds have to be in the same project? One user might have access to some projects but not all.

What do you mean by that? Seeds are global resources, every user/every project can see/use them if they are not protected.

> Would the user be allowed to move the shoot control plane to any seed they have access to, checking the seed taint seed.gardener.cloud/protected?

Probably the user will be allowed, yes. We were thinking about introducing a `/schedule` subresource to shoots so that we can RBAC-wise limit users from arbitrarily changing the seed (because they shouldn't care), but this will only be an improvement later on and not part of the initial implementation.
In the `ShootValidator` admission plugin it is ensured that no protected seed can be used, see https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/validator/admission.go#L215-L219.

> Seed and shoot networks must be disjoint. What if the destination seed has a network that overlaps with the shoot's network?

This is also prevented by the `ShootValidator` admission plugin, see https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/validator/admission.go#L371-L380. Such an update would be rejected.

Reply:
> What do you mean by that? Seeds are global resources, every user/every project can see/use them if they are not protected.

You're right, I forgot that seeds were global.

> Probably the user will be allowed, yes. We were thinking about introducing a /schedule subresource to shoots so that we can RBAC-wise limit users from arbitrarily changing the seed (because they shouldn't care), but this will only be an improvement later on and not part of the initial implementation.

Interesting. I didn't know that CRD could have custom definitions of subresources which can be mapped to specific fields in the main resource. Sounds good.

> In the ShootValidator admission plugin it is ensured that no protected seed can be used, see https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/validator/admission.go#L215-L219.

It makes sense. Thanks for the pointers!

Reply:
> Interesting. I didn't know that CRD could have custom definitions of subresources which can be mapped to specific fields in the main resource. Sounds good.

CRD cannot have it I think, but the `Shoot` (and other resources managed by Gardener) are not CRDs. They are resources served by an extension API server (gardener-apiserver), hence, you have full control and can introduce such custom subresources.